GVFS/GVFS.Platform.Mac/MacFileBasedLock.cs

```diff
@@ -0,0 +1,119 @@
+
+        private const int LockSh = 1; // #define LOCK_SH   1    /* shared lock */
+        private const int LockEx = 2; // #define LOCK_EX   2    /* exclusive lock */
+        private const int LockNb = 4; // #define LOCK_NB   4    /* don't block when locking */
+        private const int LockUn = 8; // #define LOCK_UN   8    /* unlock */
+
+        int lockFileDescriptor;
+        bool lockAcquired;
+
+        public MacFileBasedLock(
+            PhysicalFileSystem fileSystem,
+            ITracer tracer,
+            string lockPath,
+            string signature)
+            : base (fileSystem, tracer, lockPath, signature) 
+        {
+            this.lockFileDescriptor = InvalidFileDescriptor;
+        }
+
+        public override bool TryAcquireLock()
+        {
+            if (this.lockAcquired)
+            {
+                return true;
+            }
+
+            if (this.lockFileDescriptor == InvalidFileDescriptor)
+            {
+                this.lockFileDescriptor = Creat(this.LockPath, Convert.ToInt32("644", 8));
+                if (this.lockFileDescriptor == InvalidFileDescriptor)
+                {
+                    int errno = Marshal.GetLastWin32Error();
+
+                    this.Tracer.RelatedWarning($"Failed to create lock file descriptor for '{this.LockPath}': {errno}"); 
+
+                    return false;
+                }
+            }
+
+            if (Flock(this.lockFileDescriptor, LockEx | LockNb) != 0)
+            {
+                int errno = Marshal.GetLastWin32Error();
+
+                // Log error if not EWOULDBLOCK
+                this.Tracer.RelatedInfo($"Failed to acquire lock for '{this.LockPath}': {errno}");
+
+                return false;
+            }
+
+            byte[] signatureBytes = Encoding.UTF8.GetBytes(this.Signature);
+            long bytesWritten = Write(
+                this.lockFileDescriptor,
+                signatureBytes,
+                Convert.ToUInt64(signatureBytes.Length));
+            
+            if (bytesWritten == -1)
+            {
+                int errno = Marshal.GetLastWin32Error();
+
+                this.Tracer.RelatedWarning($"Failed to write signature for '{this.LockPath}': {errno}");
+            }
+
+            this.Tracer.RelatedInfo($"Lock acquired for for '{this.LockPath}'");
+
+            return true;
+        }
+
+        public override void Dispose()
+        {
+            if (this.lockAcquired)
+            {
+                if (Flock(this.lockFileDescriptor, LockUn) != 0)
+                {
+                    this.Tracer.RelatedWarning($"Failed to release lock for: '{this.LockPath}'");
+                }
+
+                this.Tracer.RelatedInfo($"Lock released: '{this.LockPath}'");
+
+                this.lockAcquired = false;
+            }
+
+            if (this.lockFileDescriptor != InvalidFileDescriptor)
+            {
+                if (Close(this.lockFileDescriptor) != 0)

```

Comment:
@sanoursa note that with the current changes the lock file is not removed from the file system.  This allows us to avoid the race conditions described here:

https://stackoverflow.com/questions/17708885/flock-removing-locked-file-without-race-condition
https://stackoverflow.com/questions/33166460/how-to-delete-a-locked-flock-file-without-race-condition-before-or-after-rele

It appears that there *are* some options for removing it if we play some tricks with `stat`, but unless you see a major issue with this approach I'm thinking we stick with this simpler approach and avoid the complexity\races of attempting to delete this file.

Note that if we were not cross platform we could `flock` the file we're interested in itself (e.g. `mapping.dat`) but to keep the cross-platform code simple and consistent it's easiest if we use a separate file on Mac like we do on Windows.